File: Java/Leetcode/Tree/SubTree.java
package com.brianprojects.Leetcode.Tree;

public class SubTree {
    //**** Subtree of Another Tree
    /*
    Given the roots of two binary trees root and subRoot, return true if there is a subtree of root
    with the same structure and node values of subRoot and false otherwise.
    A subtree of a binary tree tree is a tree that consists of a
    node in tree and all of this node's descendants. The tree tree could also be considered as a subtree of itself.
     */

    public static boolean isSubTree(TreeNode root, TreeNode subRoot) {
        if(root == null){
            return false;
        }
        if(isSameTree(root, subRoot)){
            return true;
        }

        return isSameTree(root.left, subRoot) || isSameTree(root.right, subRoot);

    }

    public static boolean isSameTree(TreeNode p, TreeNode q) {
        if(p == null && q==null){
            return true;
        }

        if(p == null && q != null || p != null && q == null || p.val != q.val){
            return false;
        }

        return (isSameTree(p.left ,q.left) && isSameTree(p.right, q.right));
    }

}

File: Python/Leetcode_Python/Trees/Subtree.py
# Time Complexity: O(n)i || Space Complexity: O(n)
from typing import Optional

from Trees.Treenode import TreeNode


def isSubTree(root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
    if not root and not subRoot:
        return True
    if (root and not subRoot) or (not root and subRoot):
        return False
    queue = []
    queue.append(root)

    while not queue:
        curr = queue.pop()
        if curr.val == subRoot.val and dfs(curr, subRoot):
            return True

        if curr.left:
            queue.append(curr.left)
        if curr.right:
            queue.append(curr.right)
    return False


def dfs(root: Optional[TreeNode], subRoot: Optional[TreeNode]):
    if not root and not subRoot:
        return True
    if (root and not subRoot) or (not root and subRoot) or root.val != subRoot.val:
        return False
    return dfs(root.left, subRoot.left) and dfs(root.right, subRoot.right)
